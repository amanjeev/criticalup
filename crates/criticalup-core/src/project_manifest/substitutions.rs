use crate::errors::ProjectManifestLoadingError;

const VARIABLE_START: &str = "${";
const VARIABLE_END: &str = "}";

enum ParseState {
    Raw,
    Variable,
}

pub(super) fn apply_substitutions(mut input: &str) -> Result<String, ProjectManifestLoadingError> {
    let mut state = ParseState::Raw;
    let mut result = String::new();

    loop {
        match state {
            ParseState::Raw => {
                if let Some(start) = input.find(VARIABLE_START) {
                    result.push_str(&input[..start]);

                    input = &input[(start + VARIABLE_START.len())..];
                    state = ParseState::Variable;
                } else {
                    // End of the input
                    result.push_str(input);
                    return Ok(result);
                }
            }
            ParseState::Variable => {
                if let Some(end) = input.find(VARIABLE_END) {
                    result.push_str(&apply_substitution(&input[..end])?);

                    input = &input[(end + VARIABLE_END.len())..];
                    state = ParseState::Raw;
                } else {
                    // End of the input
                    return Err(ProjectManifestLoadingError::UnterminatedVariableInSubstitution);
                }
            }
        }
    }
}

fn apply_substitution(variable: &str) -> Result<String, ProjectManifestLoadingError> {
    match variable {
        "rustc-host" => Ok(env!("TARGET").into()),
        other => Err(ProjectManifestLoadingError::UnknownVariableInSubstitution(
            other.into(),
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_apply_substitutions() {
        assert_eq!("hello world", apply_substitutions("hello world").unwrap());
        assert_eq!(
            env!("TARGET"),
            apply_substitutions("${rustc-host}").unwrap()
        );
        assert_eq!(
            concat!("hello ", env!("TARGET")),
            apply_substitutions("hello ${rustc-host}").unwrap()
        );
        assert_eq!(
            concat!("hello ", env!("TARGET"), "!"),
            apply_substitutions("hello ${rustc-host}!").unwrap()
        );
        assert_eq!(
            concat!("hello ", env!("TARGET"), "}"),
            apply_substitutions("hello ${rustc-host}}").unwrap()
        );

        assert!(matches!(
            apply_substitutions("hello ${").unwrap_err(),
            ProjectManifestLoadingError::UnterminatedVariableInSubstitution
        ));
        assert!(matches!(
            apply_substitutions("hello ${missing-var}!").unwrap_err(),
            ProjectManifestLoadingError::UnknownVariableInSubstitution(s) if s == "missing-var"
        ));
        assert!(matches!(
            apply_substitutions("hello ${}!").unwrap_err(),
            ProjectManifestLoadingError::UnknownVariableInSubstitution(s) if s.is_empty()
        ));
    }

    #[test]
    fn test_apply_substitution() {
        assert_eq!(env!("TARGET"), apply_substitution("rustc-host").unwrap());

        assert!(matches!(
            apply_substitution("rustc_host").unwrap_err(),
            ProjectManifestLoadingError::UnknownVariableInSubstitution(s) if s == "rustc_host"
        ));
        assert!(matches!(
            apply_substitution("").unwrap_err(),
            ProjectManifestLoadingError::UnknownVariableInSubstitution(s) if s.is_empty()
        ));
    }
}
