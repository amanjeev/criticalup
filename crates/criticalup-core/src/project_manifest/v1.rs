use serde::Deserialize;
use std::collections::HashMap;

#[derive(Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub(super) struct ProjectManifest {
    #[allow(unused)]
    manifest_version: u32,
    #[serde(default)]
    pub(super) products: HashMap<String, ProjectManifestProduct>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub(super) struct ProjectManifestProduct {
    pub(super) release: String,
    pub(super) packages: Vec<String>,
}
