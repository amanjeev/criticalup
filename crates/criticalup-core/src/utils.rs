// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::errors::WriteFileError;
use sha2::{Digest, Sha256};
use std::hash::Hasher;
use std::path::Path;
use tokio::fs::File;
use tokio::io::BufWriter;

pub(crate) async fn open_file_for_write(path: &Path) -> Result<BufWriter<File>, WriteFileError> {
    // Ensure the parent directory is always present
    if let Some(parent) = path.parent() {
        tokio::fs::create_dir_all(parent)
            .await
            .map_err(WriteFileError::CantCreateParentDirectory)?;
    }

    Ok(BufWriter::new(
        File::create(path).await.map_err(WriteFileError::Io)?,
    ))
}

/// A `Hasher` helper type which is a wrapper to a choice of cryptographic hashing algorithm
/// to generate cryptographic hash of our types. This is needed to make sure we
/// 1. do not use [`DefaultHasher`], which may change its algorithm, for hash state
/// 2. bridge the gap between normal [`Hash`] and cryptographic hash (e.g. [`Sha256`])
/// 3. better ergonomics to create a hash of our types like [`ProjectManifestProduct`] using `#[derive(Hash)]`
pub struct Sha256Hasher(Sha256);

impl Sha256Hasher {
    pub(crate) fn new() -> Self {
        Self(Sha256::new())
    }

    /// Provides the final hash value
    pub(crate) fn finalize(self) -> String {
        format!("{:x}", self.0.finalize())
    }
}

impl Hasher for Sha256Hasher {
    /// This method is unreachable and here to appease the compiler, mandatory method.
    fn finish(&self) -> u64 {
        unreachable!()
    }

    /// Update the hasher state, mandatory method.
    fn write(&mut self, bytes: &[u8]) {
        self.0.update(bytes)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::hash::Hash;

    #[test]
    fn test_sha256_for_a_struct_works() {
        #[derive(Hash)]
        struct Abc {
            name: String,
            version: String,
            dependencies: Vec<String>,
        }

        let abc = Abc {
            name: "abc".to_string(),
            version: "1.2.3".to_string(),
            dependencies: vec!["dep2".to_string(), "dep1".to_string()],
        };

        let mut hasher = Sha256Hasher::new();
        abc.hash(&mut hasher);
        let final_hash = hasher.finalize();
        assert_eq!(
            "fb9eee112b5cee551f7a5088402e53dbbdaec2a1e1cd3f4663a1b81b1b53015f",
            final_hash
        )
    }
}
