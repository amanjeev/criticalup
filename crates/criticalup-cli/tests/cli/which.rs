// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::assert_output;
use crate::utils::{construct_toolchains_product_path, TestEnvironment};
use criticalup_core::project_manifest::ProjectManifest;
use std::fs::File;

#[test]
fn help_message() {
    let test_env = TestEnvironment::prepare();
    assert_output!(test_env.cmd().args(["which", "--help"]));
}

#[test]
fn which_run_binary_exists() {
    let test_env = TestEnvironment::prepare();

    let mut current_dir =
        std::env::current_dir().expect("could not read current directory in the test.");
    current_dir.push("tests/resources/criticalup-which.toml");

    let manifest_path = current_dir.to_str().expect("conversion to str failed");

    // generate the manifest object so we can get the installation id hash
    let p = ProjectManifest::load(current_dir.as_path()).expect("could not load project manifest");
    let id_hash = p.products()[0].installation_id().0;

    // manually create the toolchain directory which allows us to skip installation
    // TODO: when tests for `install` command are up, use that instead of manual creation
    let product_toolchain_dir = construct_toolchains_product_path(&test_env, id_hash.as_str());
    let product_toolchain_bin_dir = product_toolchain_dir.join("bin");
    std::fs::create_dir_all(&product_toolchain_bin_dir)
        .expect("could not create product directory");

    // create a file "rustc" in the toolchain/.../bin
    let _ = File::create(product_toolchain_bin_dir.join("rustc")).unwrap();

    assert_output!(test_env
        .cmd()
        .args(["which", "rustc", "--project", manifest_path]));
}

#[test]
fn which_run_binary_does_not_exists() {
    let test_env = TestEnvironment::prepare();

    let mut current_dir =
        std::env::current_dir().expect("could not read current directory in the test.");
    current_dir.push("tests/resources/criticalup-which.toml");

    let manifest_path = current_dir.to_str().expect("conversion to str failed");

    // generate the manifest object so we can get the installation id hash
    let p = ProjectManifest::load(current_dir.as_path()).expect("could not load project manifest");
    let id_hash = p.products()[0].installation_id().0;

    // manually create the toolchain directory which allows us to skip installation
    // TODO: when tests for `install` command are up, use that instead of manual creation
    let product_toolchain_dir = construct_toolchains_product_path(&test_env, id_hash.as_str());
    let product_toolchain_bin_dir = product_toolchain_dir.join("bin");
    std::fs::create_dir_all(product_toolchain_bin_dir).expect("could not create product directory");

    assert_output!(test_env
        .cmd()
        .args(["which", "rustc", "--project", manifest_path]));
}
