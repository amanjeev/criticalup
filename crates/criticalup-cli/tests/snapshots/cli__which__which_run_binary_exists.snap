---
source: crates/criticalup-cli/tests/cli/which.rs
expression: repr
---
exit: exit status: 0

stdout
------
/path/to/toolchain/installation/eee0c78b0f09ca88069fa1b14c35a6e70ffbfcce50f6ffb1e567f26a76ff7e89/bin/rustc

------

empty stderr
